# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [1]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [ ]:
spaceship=spaceship.dropna()
# Extraer la letra antes de la primera barra usando regex
spaceship['Cabin'] = spaceship['Cabin'].str.extract(r'^([A-Z])')
#Se generan dataframes para características y targets
target=spaceship["Transported"]
spaceship= spaceship.drop(columns=["PassengerId", "Name", "Transported"])
features = pd.get_dummies(spaceship, columns=["HomePlanet","Cabin", "Destination"])


**Perform Train Test Split**

In [5]:


#Se ejecuta la división de datos para entrenar y probar el modelo, considerando X para features (variables independientes)
#  y Y para target (variable dependiente a predecir). Se asigna el 80% de los datos para entrenar el modelo, 
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size = 0.20, random_state=42)

# Scaling

In [6]:
# scaling the data 
from sklearn.preprocessing import MinMaxScaler, StandardScaler

scaler= StandardScaler()

scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Normalizer


In [7]:
normalizer = MinMaxScaler()
X_train_norm = normalizer.fit_transform(X_train)
X_test_norm = normalizer.transform(X_test)

In [15]:
X_train_norm = pd.DataFrame(X_train_norm, columns = features.columns)
X_train_norm.head()

CryoSleep       Age  VIP  RoomService  FoodCourt  ShoppingMall       Spa  \
0        0.0  0.316456  0.0          0.0   0.056116      0.000000  0.028650   
1        0.0  0.455696  0.0          0.0   0.088015      0.135232  0.124911   
2        1.0  0.430380  0.0          0.0   0.000000      0.000000  0.000000   
3        1.0  0.468354  0.0          0.0   0.000000      0.000000  0.000000   
4        1.0  0.278481  0.0          0.0   0.000000      0.000000  0.000000   

     VRDeck  HomePlanet_Earth  HomePlanet_Europa  ...  Cabin_B  Cabin_C  \
0  0.030094               0.0                1.0  ...      0.0      0.0   
1  0.000049               0.0                1.0  ...      0.0      1.0   
2  0.000000               0.0                0.0  ...      0.0      0.0   
3  0.000000               1.0                0.0  ...      0.0      0.0   
4  0.000000               1.0                0.0  ...      0.0      0.0   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0      0.0      0.0      0.0      0.0      0.0                      1.0   
1      0.0      0.0      0.0      0.0      0.0                      0.0   
2      0.0      0.0      1.0      0.0      0.0                      1.0   
3      0.0      0.0      0.0      1.0      0.0                      1.0   
4      0.0      0.0      0.0      1.0      0.0                      0.0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  
0                        0.0                      0.0  
1                        0.0                      1.0  
2                        0.0                      0.0  
3                        0.0                      0.0  
4                        1.0                      0.0  

[5 rows x 22 columns]

In [16]:
X_test_norm = pd.DataFrame(X_test_norm, columns = features.columns)
X_test_norm.head()

CryoSleep       Age  VIP  RoomService  FoodCourt  ShoppingMall       Spa  \
0        1.0  0.367089  0.0     0.000000   0.000000      0.000000  0.000000   
1        1.0  0.164557  0.0     0.000000   0.000000      0.000000  0.000000   
2        1.0  0.632911  0.0     0.000000   0.000000      0.000000  0.000000   
3        0.0  0.075949  0.0     0.000000   0.000000      0.000000  0.000000   
4        0.0  0.468354  0.0     0.024458   0.000101      0.049049  0.000045   

     VRDeck  HomePlanet_Earth  HomePlanet_Europa  ...  Cabin_B  Cabin_C  \
0  0.000000               0.0                1.0  ...      1.0      0.0   
1  0.000000               1.0                0.0  ...      0.0      0.0   
2  0.000000               1.0                0.0  ...      0.0      0.0   
3  0.000000               1.0                0.0  ...      0.0      0.0   
4  0.003344               1.0                0.0  ...      0.0      0.0   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0      0.0      0.0      0.0      0.0      0.0                      0.0   
1      0.0      0.0      0.0      1.0      0.0                      1.0   
2      0.0      0.0      0.0      1.0      0.0                      0.0   
3      0.0      0.0      0.0      1.0      0.0                      1.0   
4      0.0      0.0      0.0      1.0      0.0                      0.0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  
0                        0.0                      1.0  
1                        0.0                      0.0  
2                        0.0                      1.0  
3                        0.0                      0.0  
4                        0.0                      1.0  

[5 rows x 22 columns]

Para poder analizar con las variables que se hayan correlacionadas entre sí, se aplica análisis de componentes principales PCA, para unificar en un único vector de componentes principales, conservando así la mayor parte de la información (varianza) posible.
Se realizarán 5 PCA para las variables

In [8]:
#PCA para reducir la dimnesión de destination 'Destination_55 Cancri e', 'Destination_PSO J318.5-22','Destination_TRAPPIST-1e'
from sklearn.decomposition import PCA
pca_train_Destination=PCA(n_components=1)
pca_train_Destination.fit(X_train[['Destination_55 Cancri e', 'Destination_PSO J318.5-22','Destination_TRAPPIST-1e']])
X_train['Destination_PCA']=pca_train_Destination.transform(X_train[['Destination_55 Cancri e', 'Destination_PSO J318.5-22','Destination_TRAPPIST-1e']])
X_test['Destination_PCA']=pca_train_Destination.transform(X_test[['Destination_55 Cancri e', 'Destination_PSO J318.5-22','Destination_TRAPPIST-1e']])
#spaceship_2=spaceship_2.drop(columns=['Destination_55 Cancri e', 'Destination_PSO J318.5-22','Destination_TRAPPIST-1e'])

In [9]:
#PCA para reducir la dimnesión de 'HomePlanet_Earth', 'HomePlanet_Europa','HomePlanet_Mars'
columns=['Cabin_A','Cabin_B', 'Cabin_C', 'Cabin_D', 'Cabin_E', 'Cabin_F', 'Cabin_G','Cabin_T']
from sklearn.decomposition import PCA
pca_train_planet=PCA(n_components=0.7)
pca_train_planet.fit(X_train[columns])
pca_result_train=pca_train_planet.transform(X_train[columns])

num_components = pca_result_train.shape[1]

# Crear nombres de columnas genéricos basados en el número de componentes
pca_column_names = [f'PCA_cabin_{i+1}' for i in range(num_components)]

# Agregar las nuevas columnas al DataFrame X_train
X_train[pca_column_names] = pca_result_train

pca_result_test=pca_train_planet.transform(X_test[columns])

num_components = pca_result_test.shape[1]

# Crear nombres de columnas genéricos basados en el número de componentes
pca_column_names = [f'PCA_cabin_{i+1}' for i in range(num_components)]

# Agregar las nuevas columnas al DataFrame X_train
X_test[pca_column_names] = pca_result_test

In [10]:
#PCA para reducir la dimnesión de 'HomePlanet_Earth', 'HomePlanet_Europa','HomePlanet_Mars'
columns=['HomePlanet_Earth', 'HomePlanet_Europa','HomePlanet_Mars']
from sklearn.decomposition import PCA
pca_train_planet=PCA(n_components=1)
pca_train_planet.fit(X_train[columns])
X_train['homeplanet_PCA']=pca_train_planet.transform(X_train[columns])
X_test['homeplanet_PCA']=pca_train_planet.transform(X_test[columns])

In [11]:
X_train.columns

Index(['CryoSleep', 'Age', 'VIP', 'RoomService', 'FoodCourt', 'ShoppingMall',
       'Spa', 'VRDeck', 'HomePlanet_Earth', 'HomePlanet_Europa',
       'HomePlanet_Mars', 'Cabin_A', 'Cabin_B', 'Cabin_C', 'Cabin_D',
       'Cabin_E', 'Cabin_F', 'Cabin_G', 'Cabin_T', 'Destination_55 Cancri e',
       'Destination_PSO J318.5-22', 'Destination_TRAPPIST-1e',
       'Destination_PCA', 'PCA_cabin_1', 'PCA_cabin_2', 'PCA_cabin_3',
       'homeplanet_PCA'],
      dtype='object')

In [12]:
features_selected=['CryoSleep', 'Age', 'VIP', 'RoomService', 'FoodCourt', 'ShoppingMall',
       'Spa', 'VRDeck','Destination_PCA', 'PCA_cabin_1', 'PCA_cabin_2', 'PCA_cabin_3',
       'homeplanet_PCA']

In [25]:
features.shape

(6606, 22)

**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

- Bagging and Pasting

In [67]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier,AdaBoostClassifier, GradientBoostingClassifier

from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [83]:
bagging_cla = BaggingClassifier(DecisionTreeClassifier(max_depth=10),n_estimators=100, random_state=42)

In [84]:
bagging_cla.fit(X_train_norm, y_train)

BaggingClassifier(estimator=DecisionTreeClassifier(max_depth=10),
                  n_estimators=100, random_state=42)

In [85]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Hacer predicciones
pred = bagging_cla.predict(X_test_norm)

# Calcular métricas de clasificación
print("Accuracy:", accuracy_score(y_test, pred))
print("\nClassification Report:\n", classification_report(y_test, pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, pred))

Accuracy: 0.8063540090771558

Classification Report:
               precision    recall  f1-score   support

       False       0.81      0.80      0.80       653
        True       0.81      0.81      0.81       669

    accuracy                           0.81      1322
   macro avg       0.81      0.81      0.81      1322
weighted avg       0.81      0.81      0.81      1322


Confusion Matrix:
 [[522 131]
 [125 544]]


In [42]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Initialize the Pasting model
pasting_cla = BaggingClassifier(DecisionTreeClassifier(max_depth=10),n_estimators=50,
  bootstrap=False  # This ensures that no bootstrap is done, which is characteristic of Pasting
)

# Train the model with normalized data
pasting_cla.fit(X_train_norm, y_train)

# Evaluate the model's performance
predpast = pasting_cla.predict(X_test_norm)

# Calcular métricas de clasificación
print("Accuracy:", accuracy_score(y_test, predpast))
print("\nClassification Report:\n", classification_report(y_test, predpast))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, predpast))

Accuracy: 0.7881996974281392

Classification Report:
               precision    recall  f1-score   support

       False       0.81      0.74      0.78       653
        True       0.77      0.84      0.80       669

    accuracy                           0.79      1322
   macro avg       0.79      0.79      0.79      1322
weighted avg       0.79      0.79      0.79      1322


Confusion Matrix:
 [[483 170]
 [110 559]]


- Random Forests

In [86]:
forest = RandomForestClassifier(n_estimators=100,
                             max_depth=10)

In [87]:
forest.fit(X_train_norm, y_train)

RandomForestClassifier(max_depth=10)

In [88]:
predfores = forest.predict(X_test_norm)

In [89]:
# Calcular métricas de clasificación
print("Accuracy:", accuracy_score(y_test, predfores))
print("\nClassification Report:\n", classification_report(y_test, predfores))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, predfores))

Accuracy: 0.81089258698941

Classification Report:
               precision    recall  f1-score   support

       False       0.81      0.81      0.81       653
        True       0.81      0.81      0.81       669

    accuracy                           0.81      1322
   macro avg       0.81      0.81      0.81      1322
weighted avg       0.81      0.81      0.81      1322


Confusion Matrix:
 [[528 125]
 [125 544]]


- Gradient Boosting

In [59]:
gb_cla = GradientBoostingClassifier(max_depth=20,
                                   n_estimators=100)

In [60]:
gb_cla.fit(X_train_norm, y_train)

GradientBoostingClassifier(max_depth=20)

In [62]:
predgb = gb_cla.predict(X_test_norm)
# Calcular métricas de clasificación
print("Accuracy:", accuracy_score(y_test, predgb))
print("\nClassification Report:\n", classification_report(y_test, predgb))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, predgb))

Accuracy: 0.7972768532526475

Classification Report:
               precision    recall  f1-score   support

       False       0.80      0.79      0.79       653
        True       0.80      0.81      0.80       669

    accuracy                           0.80      1322
   macro avg       0.80      0.80      0.80      1322
weighted avg       0.80      0.80      0.80      1322


Confusion Matrix:
 [[514 139]
 [129 540]]


- Adaptive Boosting

In [53]:
ada_cla = AdaBoostClassifier(DecisionTreeClassifier(max_depth=20),
                            n_estimators=200)

In [54]:
ada_cla.fit(X_train_norm, y_train)

c:\Users\est_c\anaconda3\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


AdaBoostClassifier(estimator=DecisionTreeClassifier(max_depth=20),
                   n_estimators=200)

In [57]:
predada = ada_cla.predict(X_test_norm)
# Calcular métricas de clasificación
print("Accuracy:", accuracy_score(y_test,predada))
print("\nClassification Report:\n", classification_report(y_test,predada))
print("\nConfusion Matrix:\n", confusion_matrix(y_test,predada))


Accuracy: 0.7738275340393344

Classification Report:
               precision    recall  f1-score   support

       False       0.77      0.76      0.77       653
        True       0.77      0.78      0.78       669

    accuracy                           0.77      1322
   macro avg       0.77      0.77      0.77      1322
weighted avg       0.77      0.77      0.77      1322


Confusion Matrix:
 [[499 154]
 [145 524]]


Which model is the best and why?

El modelo de Random Forest tiene un rendimiento sólido con un 81% de precisión, manejando ambas clases de manera equilibrada (precision, recall y F1-score iguales). Comete 125 falsos positivos y 125 falsos negativos, mostrando una tasa de error balanceada. Para mejorar, puedes ajustar hiperparámetros como n_estimators y max_depth o usar técnicas como Grid Search o Random Search.